cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -561,151 +536,273 @@ std::vector<std::unique_ptr<column>> get_unique_entries_and_list_offsets(
-  // and the prefix sums of empty lists are { 0, 1, 1, 2, 2, 3 },
-  // then output = { 0, 0, 4, 4, 5, 5 }.
-  thrust::transform(rmm::exec_policy(stream),
-                    thrust::make_counting_iterator<offset_type>(0),
-                    thrust::make_counting_iterator<offset_type>(original_offsets.size()),
-                    original_offsets.begin<offset_type>(),
-                    [prefix_sum_empty_lists = original_offsets.begin<offset_type>(),
-                     offsets = new_offsets->view().begin<offset_type>()] __device__(auto i) {
-                      return offsets[i - prefix_sum_empty_lists[i]];
-                    });
+  // Let consider an example:
+  // Given the original offsets of the input lists column is [0, 4, 5, 6, 7, 10, 11, 13].
+  // The original entries_list_indices is [1, 1, 1, 1, 2, 3, 4, 5, 5, 5, 6, 7, 7], and after
+  // extracting unique entries we have the entries_list_indices becomes [1, 1, 1, 4, 5, 5, 5, 7, 7]
+  // and num_entries is 9. These are the input to this function.
+  //
+  // Through extracing unique list entries, one entry in the list index 1 has been removed (first
+  // list, as we are using 1-based list index), and entries in the lists with indices {3, 3, 6} have
+  // been removed completely.
+
+  // This variable stores the (1-based) list indices of the unique entries but only one index value
+  // per non-empty list. Given the example above, we will have this array hold the values
+  // [1, 4, 5, 7].
+  auto list_indices = rmm::device_uvector<size_type>(original_offsets.size() - 1, stream);
+
+  // Stores the non-zero numbers of unique entries per list.
+  // Given the example above, we will have this array contains the values [3, 1, 3, 2]
+  auto list_sizes = rmm::device_uvector<size_type>(original_offsets.size() - 1, stream);
+
+  // Count the numbers of unique entries for each non-empty list.
+  auto const end                 = thrust::reduce_by_key(rmm::exec_policy(stream),
+                                         entries_list_indices.template begin<size_type>(),
+                                         entries_list_indices.template end<size_type>(),
+                                         thrust::make_constant_iterator<size_type>(1),
+                                         list_indices.begin(),
+                                         list_sizes.begin());
+  auto const num_non_empty_lists = thrust::distance(list_indices.begin(), end.first);
+
+  // The output offsets for the output lists column(s).
+  auto new_offsets         = make_numeric_column(data_type{type_to_id<offset_type>()},
+                                         original_offsets.size(),
+                                         mask_state::UNALLOCATED,
+                                         stream,
+                                         mr);
+  auto const d_new_offsets = new_offsets->mutable_view().template begin<offset_type>();
```

Comment:
I think this would be cleaner to have `new_offsets` be a `device_uvector<size_type>` and then just use the `column(device_uvector)` ctor at the end.